src/CodeGen_Hexagon.cpp

```diff
@@ -0,0 +1,1379 @@
+            return false;
+        }
+    }
+
+    return true;
+}
+
+}  // namespace
+
+Value *CodeGen_Hexagon::shuffle_vectors(Value *a, Value *b,
+                                        const vector<int> &indices) {
+    llvm::Type *a_ty = a->getType();
+    llvm::Type *b_ty = b->getType();
+    internal_assert(a_ty == b_ty);
+
+    bool B128 = target.has_feature(Halide::Target::HVX_128);
+    int a_elements = static_cast<int>(a_ty->getVectorNumElements());
+    int b_elements = static_cast<int>(b_ty->getVectorNumElements());
+
+    llvm::Type *element_ty = a->getType()->getVectorElementType();
+    int element_bits = element_ty->getScalarSizeInBits();
+    int native_elements = native_vector_bits() / element_bits;
+    llvm::Type *native_ty = llvm::VectorType::get(element_ty, native_elements);
+    llvm::Type *native2_ty = llvm::VectorType::get(element_ty, native_elements*2);
+
+    int result_elements = static_cast<int>(indices.size());
+    llvm::Type *result_ty = VectorType::get(element_ty, result_elements);
+
+    // Try to rewrite shuffles that only access the elements of b.
+    int min = indices[0];
+    for (size_t i = 1; i < indices.size(); i++) {
+        if (indices[i] != -1 && indices[i] < min) {
+            min = indices[i];
+        }
+    }
+    if (min >= a_elements) {
+        vector<int> shifted_indices(indices);
+        for (int &i : shifted_indices) {
+            if (i != -1) i -= a_elements;
+        }
+        return shuffle_vectors(b, UndefValue::get(b->getType()), shifted_indices);
+    }
+
+    // Try to rewrite shuffles that only access the elements of a.
+    int max = *std::max_element(indices.begin(), indices.end());
+    if (max < a_elements) {
+        BitCastInst *a_cast = dyn_cast<BitCastInst>(a);
+        CallInst *a_call = dyn_cast<CallInst>(a_cast ? a_cast->getOperand(0) : a);
+        llvm::Function *vcombine =
+            Intrinsic::getDeclaration(module.get(), IPICK(Intrinsic::hexagon_V6_vcombine));
+        if (a_call && a_call->getCalledFunction() == vcombine) {
+            // Rewrite shuffle(vcombine(a, b), x) to shuffle(a, b)
+            return shuffle_vectors(
+                create_bitcast(a_call->getArgOperand(1), native_ty),
+                create_bitcast(a_call->getArgOperand(0), native_ty),
+                indices);
+        } else if (ShuffleVectorInst *a_shuffle = dyn_cast<ShuffleVectorInst>(a)) {
+            bool is_identity = true;
+            for (int i = 0; i < a_elements; i++) {
+                int mask_i = a_shuffle->getMaskValue(i);
+                is_identity = is_identity && (mask_i == i || mask_i == -1);
+            }
+            if (is_identity) {
+                return shuffle_vectors(
+                    a_shuffle->getOperand(0),
+                    a_shuffle->getOperand(1),
+                    indices);
+            }
+        }
+    }
+
+    // Try to rewrite shuffles of (maybe strided) ramps.
+    int start, stride;
```

Comment:
is_strided_ramp() may always set these, but I'd still feel better seeing them explicitly initialized to something, just in case